src/kvstore/raftex/RaftPart.cpp

```diff
@@ -415,6 +419,82 @@ void RaftPart::commitTransLeader(const HostAddr& target) {
     }
 }
 
+void RaftPart::updateQuorum() {
+    CHECK(!raftLock_.try_lock());
+    int32_t total = 0;
+    for (auto& h : hosts_) {
+        if (!h->isLearner()) {
+            total++;
+        }
+    }
+    quorum_ = (total + 1) / 2;
+}
+
+void RaftPart::addPeer(const HostAddr& peer) {
+    CHECK(!raftLock_.try_lock());
+    if (peer == addr_) {
+        LOG(INFO) << idStr_ << "I am already in the raft group!";
+        return;
+    }
+    auto it = std::find_if(hosts_.begin(), hosts_.end(), [&peer] (const auto& h) {
+                return h->address() == peer;
+            });
+    if (it == hosts_.end()) {
+        hosts_.emplace_back(std::make_shared<Host>(peer, shared_from_this()));
+        updateQuorum();
+        LOG(INFO) << idStr_ << "Add peer " << peer;
+    } else {
+        if ((*it)->isLearner()) {
+            LOG(INFO) << idStr_ << "The host " << peer
+                      << " has been existed as learner, promote it!";
+            (*it)->setLearner(false);
+            updateQuorum();
+        } else {
+            LOG(INFO) << idStr_ << "The host " << peer << " has been existed as follower!";
+        }
+    }
+}
+
+void RaftPart::removePeer(const HostAddr& peer) {
+    CHECK(!raftLock_.try_lock());
+    LOG(INFO) << idStr_ << "Remove peer " << peer;
+    if (peer == addr_) {
+        status_ = Status::STOPPED;
+        LOG(INFO) << idStr_ << "Remove myself from the raft group, just stop.";
+        return;
+    }
+    auto it = std::find_if(hosts_.begin(), hosts_.end(), [&peer] (const auto& h) {
+                  return h->address() == peer;
+              });
+    if (it == hosts_.end()) {
+        LOG(INFO) << idStr_ << "The peer " << peer << " not exist!";
+    } else {
+        CHECK(!(*it)->isLearner()) << idStr_ << "Peer " << peer << " should not be learner!";
```

Comment:
the check mean is that we can not remove a learner from this interface?

Reply:
Good point.  Maybe we could reuse the interface to remove a learner.